Ca 291 newsletter signup

arnoutj opened the pull request:
Replaced Mailchimp form in the footer with newsletter signup React component. 
This was necessary in order to show visual feedback after signup (success or failure).

timstokman commented on CollAction/Frontend/app/global/NewsletterSignup.tsx:
```diff
@@ -0,0 +1,141 @@
+  }
+
+  isValidEmail() {
+    return /[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,3}/.test(this.state.email);
```
Validating the e-mail through a regex isn't a great idea, this will have a lot of false negatives. Maybe only if it contains the '@' symbol? https://davidcel.is/posts/stop-validating-email-addresses-with-regex/

arnoutj replied:
Yes, thanks. I will update it.

arnoutj replied:
Updated it to a check for just '@' symbol as you suggested @timstokman

timstokman commented on CollAction/Frontend/app/global/styles/footer.scss:
```diff
@@ -117,5 +89,76 @@
+        position: relative;
+    }
+
+    .newsletter-signup__container {
```
Bit of a weird naming convention. Why two underscores?

arnoutj replied:
It's the block-element-modifier (BEM) convention: https://csswizardry.com/2013/01/mindbemding-getting-your-head-round-bem-syntax/. In this case newsletter-signup is the block and container the element. 

We used this at my previous employer for all websites. It may look weird in the beginning but works quite well. I woud like to introduce this convention in the whole frontend later on. But we can discuss that.

timstokman commented on CollAction/Frontend/app/global/NewsletterSignup.tsx:
```diff
@@ -0,0 +1,141 @@
+}
+
+class NewsletterSignup extends React.Component<NewsLetterSignupProps, NewsLetterSignupState> {
+
```
Inconsistent newline

arnoutj replied:
I'll fix it

arnoutj replied:
Done

timstokman replied:
Not pushed yet?

arnoutj replied:
I don't see the problem. New line before and after class definition like in other components..

timstokman replied:
Here's a typical react example: https://reactjs.org/docs/components-and-props.html

This is just making sure the formatting is consistent with typical typescript/react code.

timstokman replied:
Would be good to add a newline between the interface definition and the class as well.

arnoutj replied:
Ok I updated it.

timstokman commented on CollAction/Frontend/app/global/NewsletterSignup.tsx:
```diff
@@ -0,0 +1,141 @@
+      if (response.ok) {
+        this.showSignupSucces();
+      } else {
+        this.showSignupError();
```
It might be the case that the http response actually has the signup error message, that might be useful to show as well (big difference between 'already subscribed', 'we suspect you might be a spammer', etc). Might be a good idea to check if that's the case.

arnoutj replied:
Fixed it.

timstokman commented:
Great work! Just had some minor comments.

You seem to have some build failures (see checks at the bottom of the pull request), might want to check that out. The latest master also has some buildfixes.

arnoutj commented:
@timstokman the build failed on a problem that was on master as well, because it remained after I rebased onto master. I fixed it in my latest commit together with your other remarks.

timstokman commented on CollAction/Frontend/app/global/NewsletterSignup.tsx:
```diff
@@ -0,0 +1,140 @@
+
+    this.setState({ subscriptionState: States.LOADING });
+
+    fetch("http://collaction.us14.list-manage.com/subscribe/post", {
```
This mailchimp account should probably be a configuration value.

timstokman commented on CollAction/Views/Shared/_Footer.cshtml:
```diff
@@ -59,35 +59,10 @@
+        <h5 class="header">@Localizer["Sign up to our newsletter"]</h5>
 
+        <!-- Mailchimp signup form -->
+        <div class="newsletter-signup" data-u="48e9b2f8f522cf59b9d5ffa8d" data-list-id="@NewsletterSubscriptionServiceOptions.Value.MailChimpNewsletterListId"></div>
```
48e9b2f8f522cf59b9d5ffa8d probably needs to be fed from a configuration value, just like @NewsletterSubscriptionServiceOptions.Value.MailChimpNewsletterListId

arnoutj replied:
I added it as a new variable @NewsletterSubscriptionServiceOptions.Value.MailChimpUserId

timstokman commented on CollAction/Services/Newsletter/NewsletterSubscriptionService.cs:
```diff
@@ -59,6 +60,7 @@ public NewsletterSubscriptionService(IOptions<NewsletterSubscriptionServiceOptio
             _newsletterListId = options.Value.MailChimpNewsletterListId;
             _jobClient = jobClient;
             _logger = logger;
+            _userId = options.Value.MailChimpUserId;
```
If it's not used in this service, it should be added as a variable.

arnoutj replied:
Do you mean as a variable in Index.cshtml?

timstokman replied:
No, I mean _userId. It looks like a new variable, and it's not used anywhere, so it's redundant.

arnoutj replied:
I see. I removed it

ljwboschker commented:
I think be can remove this PR, right?

arnoutj commented:
Yes!

> Op 7 jun. 2019 om 13:57 heeft Luuk <notifications@github.com> het volgende geschreven:
> 
> I think be can remove this PR, right?
> 
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub, or mute the thread.
